mirgecom/fluid.py

```diff
@@ -0,0 +1,65 @@
+""":mod:`mirgecom.fluid` provides common utilities for fluid simulation.
+
+.. autofunction:: compute_local_velocity_gradient
+"""
+
+__copyright__ = """
+Copyright (C) 2021 University of Illinois Board of Trustees
+"""
+
+__license__ = """
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+"""
+import numpy as np  # noqa
+from meshmode.dof_array import DOFArray  # noqa
+from mirgecom.euler import ConservedVars
+
+
+def compute_local_velocity_gradient(discr, cv: ConservedVars):
+    r"""
+    Compute the cell-local gradient of fluid velocity.
+
+    Computes the cell-local gradient of fluid velocity from:
+
+    .. math::
+
+        \nabla{v_i} = \frac{1}{\rho}(\nabla(\rho{v_i})-v_i\nabla{\rho}),
+
+    where $v_i$ is ith velocity component.
+
+    Parameters
+    ----------
+    discr: grudge.eager.EagerDGDiscretization
+        the discretization to use
+    cv: mirgecom.euler.ConservedVars
+        the fluid conserved variables
+    Returns
+    -------
+    numpy.ndarray
+        object array of :class:`~meshmode.dof_array.DOFArray`
+        representing $\partial_j{v_i}$. e.g. for 2D:
+        $\left( \begin{array}{cc}
+        \partial_{x}\mathbf{v}_{x}&\partial_{y}\mathbf{v}_{x} \\
+        \partial_{x}\mathbf{v}_{y}&\partial_{y}\mathbf{v}_{y} \end{array} \right)$
+    """
+    dim = discr.dim
+    velocity = cv.momentum/cv.mass
+    dmass = discr.grad(cv.mass)
+    dmom = np.array([discr.grad(cv.momentum[i]) for i in range(dim)], dtype=object)
+    return (dmom - np.outer(velocity, dmass))/cv.mass
```

Comment:
In addition to the missing surface part, this here also still makes me quite nervous, TBH. As it evaluates a fairly nonlinear function (division) on a non-upsampled grid. (And there's no facility here to get onto one.)

Could you cross-reference this with what is done in the DG book? If they do the same thing, it would make me less nervous.

Reply:
@inducer I don't think there needs to be any surface term here because this is just computing a term inside the viscous flux, not a weak divergence or numerical flux. I.e. this is a utility function designed to help construct the flux, which will later be integrated by parts. Then we will run into the question of how to handle the numerical flux (on element boundaries) afterward.

Reply:
@thomasgibson is correct about the intended use for this routine. However, maybe I should just remove it from this PR and then re-submit when its use makes the intent apparent.

Reply:
Updated this function (091b432) so that it is what it ended up being from NS #298.  No more sketchy _local grad_ is used internally. In practice in NS, a _dg grad_ is used as input.
   
@inducer, @thomasgibson